### PR TITLE
[WIP] Adding support for storing multiple shapes in a Polyline instance

### PR DESCRIPTION
**Pull request by brimoor**
Requires https://github.com/voxel51/eta/pull/487 in order to function.

Data formats like COCO and BDD allow for the possibility of a single "object" being composed of multiple disjoint polylines/polygons. This PR updates the data model to allow for this. IE `Polyline.points` is now a list of lists of `(x, y)` points.

The previous behavior in these cases is split between either:

(a) generating multiple `Polyline` instances for complex objects that are composed of multiple shapes, thereby losing invertibility of importing/exporting a dataset in these formats (complex objects will become separate objects in the new export)

(b) only loading the first shape and dropping others

We can discuss implications of this. Marked as `[WIP]` until the required frontend updates and possibly a DB migration are in here.

The DB migration is only necessary if we are worried that folks may have created polyline datasets within the last week, since the feature was only launched in v0.6.1.

Here's an example of one "object" with three shapes (the three red regions on the skis, which are instance masks)

### CODE REVIEW

**Comment by benjaminpkane** [comment]
I think this this should be a `0.6.3` item noting the App and migration needs.

**Comment by lethosor** [comment]
Repeating a comment from https://github.com/voxel51/eta/pull/487#issuecomment-708489664 that I meant to leave here:

> There doesn't appear to be error checking for polylines using the old format. Should there be?
> 
> ```python
> fo.Polyline(points=[(0.1, 0.1), (0.3, 0.1), (0.3, 0.3), (0.1, 0.3)])
> fo.Polyline(points=[[(0.1, 0.1), (0.3, 0.1), (0.3, 0.3), (0.1, 0.3)]])
> ```

I think some type-checking here could be beneficial to help users track down any problems caused by this, although it may not be as necessary with a migration.

**Comment by brimoor** [comment]
Alright I added some validation:

```py
import fiftyone as fo

fo.Polyline()  # okay
fo.Polyline(points=[[(1, 2)]])  # okay
fo.Polyline(points=[[(1, 2, 3)]])  # bad
fo.Polyline(points=[(1, 2)])  # bad

fo.Keypoint()  # okay
fo.Keypoint(points=[(1, 2)])  # okay
fo.Keypoint(points=[(1, 2, 3)])  # bad
fo.Keypoint(points=[[(1, 2)]])  # bad
```

In general I'm not too wild about adding validation right now because (a) MongoEngine is slow and we want to get rid of it, and (b) I'm not certain that Python-side validation makes sense even w/o MongoEngine, given that a top priority is for everything to be much faster than it is now.

**Comment by lethosor** [comment]
Works for me, and understood on the speed concerns. I would generally prefer early validation like this, since it makes it easier to locate the source of incorrectly-typed data, but maybe exempting internal code (e.g. importers/exporters) for performance reasons would be possible eventually. I'm also not too attached to validation in this particular case - if it proves to be a bottleneck and nobody has data in the previous format anymore, I think removing it would be fine.

**Comment by brimoor** [comment]
I haven't benchmarked, but this probably makes creating `Polyline` and `Keypoint` instances *faster*. For example, `Polyline.points` was previously `ListField(ListField())`, whose default validation probably insisted that every element of the outer list was a list, which I'm not doing anymore.

Working with real-life datasets like COCO-2017 with segmentations loaded, I've seen that loading samples with polylines can be quite slow. #MongoEngineMustDie

**Review by lethosor — APPROVED** [review]

**Comment by lethosor** [comment]
Actually, does this one still need a DB migration?

**Comment by benjaminpkane** [comment]
I'm adding to 

> Actually, does this one still need a DB migration?

I'm updating the revision right now. It only needs it we don't want to break persistent datasets in the wild right now that have polylines.

**Comment by benjaminpkane** [comment]
I have updated (and tested) the revision to include polyline updates and resolve #607.

File: fiftyone/migrations/revisions/v0_6_2.py [hunk 1]
```diff
@@ -21,20 +35,36 @@ def up(db, dataset_name):
...
+
     db.datasets.replace_one(match_d, dataset_dict)
+    sample_coll = dataset_dict["sample_collection_name"]
 
     if dataset_dict["media_type"] == "image":
+        writes = []
+        for s in db[sample_coll].find():
+            converted = False
+            for d in s.items():
+                converted = _up_convert_polyline_points(s)
```
Comment (lethosor) [inline fiftyone/migrations/revisions/v0_6_2.py]: This change would catch cases where only some (but not the last) instances are converted:
```suggestion
                converted |= _up_convert_polyline_points(s)
```
Same goes for the other two instances of loops like this.

File: fiftyone/migrations/revisions/v0_6_2.py [hunk 2]
```diff
@@ -8,6 +8,20 @@
 import pymongo as pm
 
 
+def _up_convert_polyline_points(d):
+    cls = d.get("cls", None)
+    if cls == "Polyline":
+        if "points" not in d:
+            return False
+        d["points"] = [d["points"]]
```
Comment (lethosor) [inline fiftyone/migrations/revisions/v0_6_2.py:18]: Is the current type of `d["points"]` checked somewhere (or does it not need to be)?

Comment (benjaminpkane) [reply fiftyone/migrations/revisions/v0_6_2.py:18]: I don't think it needs to be. Mongengine validates these fields.

**Pull request merged**